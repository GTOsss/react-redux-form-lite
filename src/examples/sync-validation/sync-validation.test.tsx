import React from 'react';
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import {createStore, applyMiddleware, combineReducers} from 'redux';
import ReduxThunkTester from 'redux-thunk-tester';
import SyncValidation from './sync-validation';
import {reducer} from '../../index';

const renderComponent = (onSubmit) => {
  const reduxThunkTester = new ReduxThunkTester();

  const store = createStore(
    combineReducers({reduxForm: reducer}),
    applyMiddleware(reduxThunkTester.createReduxThunkHistoryMiddleware()),
  );

  const component = mount(
    <Provider store={store}><SyncValidation onSubmit={onSubmit} /></Provider>,
  );

  return {reduxThunkTester, store, component};
};

describe('<SyncValidation />', () => {
  test('render sync validation form (inputs)', () => {
    const {component} = renderComponent(() => {
    });
    expect(component.find('input')).toMatchSnapshot();
  });

  test('validate only change', () => {
    const {component, store} = renderComponent(() => {
    });

    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('validate', () => {
    const {component, store} = renderComponent(() => {
    });

    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('validate: onSubmit after failed submit', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(1);
    expect(onSubmit.mock.calls[0][0].values).toEqual({firstName: 'test', lastName: undefined});
    expect(onSubmit.mock.calls[0][0].state.form.hasErrors).toBeTruthy();
    expect(onSubmit.mock.calls[0][0].state.form.hasWarnings).toBeFalsy();
    expect(onSubmit.mock.calls[0][0].state.form.errorsMap).toEqual({lastName: 'Field required.'});
    expect(onSubmit.mock.calls[0][0].state.form.warningsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('validate: onSubmit after failed and success', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(2);
    expect(onSubmit.mock.calls[1][0].values).toEqual({firstName: 'test', lastName: 'test'});
    expect(onSubmit.mock.calls[1][0].state.form.hasErrors).toBeFalsy();
    expect(onSubmit.mock.calls[1][0].state.form.hasWarnings).toBeFalsy();
    expect(onSubmit.mock.calls[1][0].state.form.errorsMap).toEqual({});
    expect(onSubmit.mock.calls[1][0].state.form.warningsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('warn only change', () => {
    const {component, store} = renderComponent(() => {
    });

    component.find('input').at(0).simulate('change', {target: {value: 't'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});

    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('warn', () => {
    const {component, store} = renderComponent(() => {
    });

    component.find('input').at(0).simulate('change', {target: {value: 't'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');

    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('warn: onSubmit after failed submit', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 't'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(1);
    expect(onSubmit.mock.calls[0][0].values).toEqual({firstName: 't', lastName: 'test'});
    expect(onSubmit.mock.calls[0][0].state.form.hasErrors).toBeFalsy();
    expect(onSubmit.mock.calls[0][0].state.form.hasWarnings).toBeTruthy();
    expect(onSubmit.mock.calls[0][0].state.form.warningsMap).toEqual({
      firstName: 'The minimum length of the value must be 2.',
    });
    expect(onSubmit.mock.calls[0][0].state.form.errorsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });

  test('warn: onSubmit after failed and success', () => {
    const onSubmit = jest.fn();
    const {component, store} = renderComponent(onSubmit);

    component.find('input').at(0).simulate('change', {target: {value: 't'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    component.find('input').at(0).simulate('change', {target: {value: 'test'}});
    component.find('input').at(1).simulate('change', {target: {value: 'test'}});
    component.find('form').simulate('submit');
    expect(onSubmit.mock.calls.length).toBe(2);
    expect(onSubmit.mock.calls[1][0].values).toEqual({firstName: 'test', lastName: 'test'});
    expect(onSubmit.mock.calls[1][0].state.form.hasErrors).toBeFalsy();
    expect(onSubmit.mock.calls[1][0].state.form.hasWarnings).toBeFalsy();
    expect(onSubmit.mock.calls[1][0].state.form.errorsMap).toEqual({});
    expect(onSubmit.mock.calls[1][0].state.form.warningsMap).toEqual({});
    expect(store.getState().reduxForm.example).toMatchSnapshot();
  });
});
