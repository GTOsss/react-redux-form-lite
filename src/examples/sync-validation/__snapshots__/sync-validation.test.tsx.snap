// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SyncValidation /> render sync validation form (inputs) 1`] = `
Array [
  <input
    name="firstName"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="First Name"
    type="text"
    value=""
  />,
  <input
    name="lastName"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Last Name"
    type="text"
    value=""
  />,
]
`;

exports[`<SyncValidation /> validate 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {
      "lastName": "Field required.",
    },
    "focused": false,
    "hasErrors": true,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": false,
      "error": "Field required.",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": undefined,
  },
}
`;

exports[`<SyncValidation /> validate only change 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {
      "lastName": "Field required.",
    },
    "focused": false,
    "hasErrors": true,
    "hasWarnings": false,
    "submitted": false,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": false,
      "error": "Field required.",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": undefined,
  },
}
`;

exports[`<SyncValidation /> validate: onSubmit after failed and success 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": "test",
  },
}
`;

exports[`<SyncValidation /> validate: onSubmit after failed submit 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {
      "lastName": "Field required.",
    },
    "focused": false,
    "hasErrors": true,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": false,
      "error": "Field required.",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": undefined,
  },
}
`;

exports[`<SyncValidation /> warn 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": true,
    "submitted": true,
    "warningsMap": Object {
      "firstName": "The minimum length of the value must be 2.",
    },
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "The minimum length of the value must be 2.",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "t",
    "lastName": "test",
  },
}
`;

exports[`<SyncValidation /> warn only change 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": true,
    "submitted": false,
    "warningsMap": Object {
      "firstName": "The minimum length of the value must be 2.",
    },
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "The minimum length of the value must be 2.",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "t",
    "lastName": "test",
  },
}
`;

exports[`<SyncValidation /> warn: onSubmit after failed and success 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": false,
    "submitted": true,
    "warningsMap": Object {},
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "test",
    "lastName": "test",
  },
}
`;

exports[`<SyncValidation /> warn: onSubmit after failed submit 1`] = `
Object {
  "form": Object {
    "activeField": "",
    "blurred": false,
    "changed": true,
    "errorsMap": Object {},
    "focused": false,
    "hasErrors": false,
    "hasWarnings": true,
    "submitted": true,
    "warningsMap": Object {
      "firstName": "The minimum length of the value must be 2.",
    },
  },
  "meta": Object {
    "firstName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "The minimum length of the value must be 2.",
    },
    "lastName": Object {
      "active": false,
      "blurred": false,
      "changed": true,
      "error": "",
      "focused": false,
      "warning": "",
    },
  },
  "values": Object {
    "firstName": "t",
    "lastName": "test",
  },
}
`;
